lib/bricks-net-api-types.js

```diff
@@ -0,0 +1,1176 @@
+			);
+		}
+		if (_this.request_body_content_type_) {
+			writes.push("Content-Type: " + _this.request_body_content_type_ + "\r\n");
+		}
+	
+		if (_this.request_has_body_) {
+			writes.push("Content-Length: " + _this.request_body_contents_.length + "\r\n");
+			writes.push("\r\n");
+			writes.push(_this.request_body_contents_);
+		}
+		else {
+			writes.push("\r\n");
+		}
+		
+		var sequence = [];
+		
+		var socket;
+		
+		function cleanup() {
+			_this._DEBUG_LOG('HTTPClient#_performRequest: Cleanup.');
+			
+			if (socket) {
+				socket.destroy();
+				socket = null;
+			}
+		}
+		
+		function finish(err) {
+			_this._DEBUG_LOG('HTTPClient#_performRequest: Finish:', err || 'OK');
+			
+			cleanup();
+			
+			if (err) { return _this.resolver.reject(err); }
+			_this.resolver.resolve(_this);
+		}
+		
+		var promise = when.promise(function (resolve, reject) {
+			_this._DEBUG_LOG('HTTPClient#_performRequest: Connect:',
+				_this.parsed_url_.host, _this.parsed_url_.port);
+			
+			socket = new net.Socket();
+			
+			socket.connect({
+				host:  _this.parsed_url_.host,
+				port:  _this.parsed_url_.port
+			}, function (err) {
+				_this._DEBUG_LOG('HTTPClient#_performRequest: Connect callback:', err || 'OK');
+				
+				if (err) { return reject(err); }
+				resolve();
+			});
+		});
+		
+		function writeNext() {
+			var chunk = writes.shift();
+			
+			return when.promise(function (resolve, reject) {
+				_this._DEBUG_LOG('HTTPClient#_performRequest: Write:', JSON.stringify(chunk));
+				
+				socket.write(chunk, 'utf8', function (err) {
+					_this._DEBUG_LOG('HTTPClient#_performRequest: Write callback:', err || 'OK');
+					
+					if (err) { return reject(err); }
+					resolve();
+				});
+			});
+		}
+		
+		writes.forEach(function () {
+			promise = promise.then(writeNext);
+		});
+		
+		promise = promise.then(function () {
+			_this._DEBUG_LOG('HTTPClient#_performRequest: new HTTPRedirectableRequestData');
+			
+			_this.http_request_ = new HTTPRedirectableRequestData(socket);
+			
+			return _this.http_request_;
+		});
+		
+		promise = promise.then(function () {
+			var response_code_as_int = parseInt(_this.http_request_.RawPath(), 10);
+			
+			_this._DEBUG_LOG('HTTPClient#_performRequest: Parsing response code:', response_code_as_int);
+			
+			_this.response_code_ = HTTPResponseCode(response_code_as_int);
+			
+			if (response_code_as_int >= 300 && response_code_as_int <= 399 && _this.http_request_.location) {
+				_this._DEBUG_LOG('HTTPClient#_performRequest: Got a redirect:', _this.http_request_.location);
+				
+				// Note: This is by no means a complete redirect implementation.
+				_this.parsed_url_ = URL(_this.http_request_.location, _this.parsed_url_);
+				_this.response_url_after_redirects_ = _this.parsed_url_.ComposeURL();
+				
+				// Close the current socket before sending the next request.
+				cleanup();
+				
+				return _this._performRequest();
+			}
+			else {
+				finish();
+			}
+		});
+		
+		return promise;
```

Comment:
This logic looks covered by the tests, except for the error conditions which I will debug later.